nomad/state/state_store.go

```diff
@@ -1136,6 +1177,109 @@ func (s *StateStore) getJobStatus(txn *memdb.Txn, job *structs.Job, evalDelete b
+	existing, err := s.JobSummary(job.ID)
+	if err != nil {
+		return fmt.Errorf("unable to retrieve summary for job: %v", err)
+	}
+	if existing == nil {
+		existing = &structs.JobSummary{
+			JobID:   job.ID,
+			Summary: make(map[string]structs.TaskGroupSummary),
+		}
+	}
+	for _, tg := range job.TaskGroups {
+		if summary, ok := existing.Summary[tg.Name]; !ok {
+			newSummary := structs.TaskGroupSummary{
+				Name:     tg.Name,
+				Queued:   tg.Count,
+				Complete: 0,
+				Failed:   0,
+				Running:  0,
+				Starting: 0,
+			}
+			existing.Summary[tg.Name] = newSummary
+		} else {
+			if summary.Queued > tg.Count {
+				summary.Queued = tg.Count
+			}
+			existing.Summary[tg.Name] = summary
+		}
+	}
+
+	if err := txn.Insert("jobsummary", existing); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *StateStore) updateSummaryWithAlloc(newAlloc *structs.Allocation,
+	existingAlloc *structs.Allocation, txn *memdb.Txn) error {
+
+	existing, err := s.JobSummary(newAlloc.JobID)
+	if err != nil {
+		return fmt.Errorf("lookup of job summary failed: %v", err)
+	}
+
+	// If we can't find an existing job summary entry then we are not going to create a
+	// new job summary entry for an allocation with that job id since we don't
+	// know the task group counts for that jobA
+	// TODO May be we can query the job and scan all the allocations for that
+	// job and create the summary before applying the change of summary state
+	// that this allocation would cause.
+	if existing == nil {
+		return nil
+	}
+
+	tgSummary, ok := existing.Summary[newAlloc.TaskGroup]
+	if !ok {
+		return nil
+	}
+	if existingAlloc == nil {
+		switch newAlloc.DesiredStatus {
+		case structs.AllocDesiredStatusFailed:
+			tgSummary.Failed += 1
+		case structs.AllocDesiredStatusStop:
+			tgSummary.Complete += 1
+		}
+		switch newAlloc.ClientStatus {
+		case structs.AllocClientStatusPending:
+			tgSummary.Starting += 1
+		case structs.AllocClientStatusRunning:
+			tgSummary.Running += 1
+		}
+		tgSummary.Queued -= 1
+	} else if existingAlloc.ClientStatus != newAlloc.ClientStatus {
+		switch newAlloc.ClientStatus {
+		case structs.AllocClientStatusRunning:
+			tgSummary.Running += 1
+		case structs.AllocClientStatusFailed:
+			tgSummary.Failed += 1
+		case structs.AllocClientStatusPending:
+			tgSummary.Starting += 1
+		case structs.AllocClientStatusComplete:
+			tgSummary.Complete += 1
+		}
+
+		switch existingAlloc.ClientStatus {
+		case structs.AllocClientStatusRunning:
+			tgSummary.Running -= 1
+		case structs.AllocClientStatusFailed:
+			tgSummary.Failed -= 1
+		case structs.AllocClientStatusPending:
+			tgSummary.Starting -= 1
+		case structs.AllocClientStatusComplete:
+			tgSummary.Complete -= 1
+		}
+	}
+
+	existing.Summary[newAlloc.TaskGroup] = tgSummary
+	if err := txn.Insert("jobsummary", existing); err != nil {
+		return fmt.Errorf("inserting job summary failed: %v", err)
+	}
+	return nil
+}
+
```

Comment:
Add restore method in StateStore and FSM.

Can see this PR: https://github.com/hashicorp/nomad/pull/540